### PR TITLE
Create 20240508-computation-limit-hike.md

### PR DESCRIPTION
**Pull request by KshitijChaudhary666**
Creating new draft FLIP

### CODE REVIEW

**Comment by KshitijChaudhary666** [comment]
Issue created here to keep the FLIP tracker updated - https://github.com/onflow/flips/issues/267

File: governance/20240508-computation-limit-hike.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+Please refer to [this post](https://forum.flow.com/t/how-evm-transaction-fees-work-on-flow-previewnet/5751) for detailed insights into how transaction fees for EVM transactions will be calculated once EVM goes live on Flow Mainnet. With the introduction of the new parameter `EVMGasUsage`, the calculation of execution effort on Flow Cadence will be modified as follows:
+
+```
+Execution Effort =
+    0.0239 * function_or_loop_call +
+    0.0123 * GetValue +
+    0.0117 * SetValue +
+		43.2994 * CreateAccount +
+	  **EVMGasUsageCost * EVMGasUsage**
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:20]:
```suggestion
	  * EVMGasUsageCost * EVMGasUsage**
```

Comment (vishalchangrani) [reply governance/20240508-computation-limit-hike.md:20]: think you are trying to bold elements within a code block..don't think that will work.

File: governance/20240508-computation-limit-hike.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# FLIP-267: Increasing Computation Limit on Flow
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:1]:
```suggestion
# FLIP-267: Increasing the transaction computation limit
```

File: governance/20240508-computation-limit-hike.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+Here, `EVMGasUsage` represents the gas used by EVM, such as 21K gas for a simple send transaction. Meanwhile, `EVMGasUsageCost` denotes the ratio for converting EVM gas to Flow’s computation units, which may also be called “Gas-to-Compute ratio” or “G2C ratio.”
+
+In this FLIP, we make two proposals -
+
+1. A mechanism to convert EVM gas costs into Flow computation costs so that EVM transactions and operations can be added into the standard Flow transaction fee calculations
+2. An increase in the computation limit for Flow transactions to reflect the performance improvements made to Cadence and the overall protocol, in order to allow each transaction to do more (when warranted).
+
+## Problem Statement
+
+EVM launch on Flow (”Crescendo”) presents an opportunity to better align with network objectives and user needs in terms of total computation required. The current execution effort (”compute”) limit of 9999 on Flow Cadence may be inadequate for larger EVM transactions. Flow should be able to match or surpass the capabilities of Ethereum, allowing transactions of up to 30M gas to be executed seamlessly.
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:32]:
```suggestion
EVM on Flow (”Crescendo”)  presents an interesting challenge - how to compute Flow computation units from EVM gas. The current execution effort (”compute”) limit of 9999 on Flow Cadence may be inadequate for larger EVM transactions. Flow should be able to match or surpass the capabilities of Ethereum, allowing transactions of up to 30M gas to be executed seamlessly.
```

File: governance/20240508-computation-limit-hike.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+To determine the appropriate reduction factor for the weights, a thorough understanding of Flow’s execution capabilities in handling large transactions was necessary. This involved assessing how much “computation” could currently be accommodated in a block (on Mainnet24) and how much additional time a transaction could consume before encountering complications. For instance, if transactions in a block exceed the available block time (1.25s), they might need to be split across multiple blocks — a feature not currently supported; also, single transactions can never be split further. After evaluating these factors, it became apparent that increasing the transaction time by more than 5 times may lead to execution issues. Therefore, it is proposed to reduce the computation weights by a factor of 5.
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:50]:
```suggestion
To determine the appropriate reduction factor for the weights, a thorough understanding of Flow’s execution capabilities in handling large transactions was necessary. This involved assessing how much “computation” could currently be accommodated in a block and how much additional time a transaction could consume before encountering complications. For instance, if transactions in a block exceed the available block time (1.25s), they might need to be split across multiple blocks — a feature not currently supported; also, single transactions can never be split further. After evaluating these factors, it became apparent that increasing the transaction time by more than 5 times may lead to execution issues. Therefore, it is proposed to reduce the computation weights by a factor of 5.
```

Comment (vishalchangrani) [reply governance/20240508-computation-limit-hike.md:50]: "before encountering complication" please can you elaborate in the text. The wording sounds very vague.

File: governance/20240508-computation-limit-hike.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+Note that the proposed reduction in coefficients is proposed to be offset by a corresponding increase in the unit cost of execution on Flow (by 5x). This adjustment aims to ensure that there are no changes in the overall transaction fee paid for a transaction despite the higher computation limit and lower coefficients. See [this post](https://forum.flow.com/t/how-evm-transaction-fees-work-on-flow-previewnet/5751) to review the calculation method for transaction fees and to understand how a corresponding increase in the cost of execution units would negate any influence on the total transaction fee.
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:52]:
```suggestion
Note that the reduction in coefficients will be offset by a corresponding increase in the unit cost of execution on Flow (by 5x). This adjustment aims to ensure that there are no changes in the overall transaction fee paid for a transaction despite the higher computation limit and lower coefficients. See [this post](https://forum.flow.com/t/how-evm-transaction-fees-work-on-flow-previewnet/5751) to review the calculation method for transaction fees and to understand how a corresponding increase in the cost of execution units would negate any influence on the total transaction fee.
```

File: governance/20240508-computation-limit-hike.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+## Proposal (Summary)
+
+1. Establish the “Gas to compute ratio” at 5000:1 by setting `EVMGasUsageCost` to `1/5000`.
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:56]:
```suggestion
1. Establish the “Gas to compute ratio” at 5000:1 by setting `EVMGasUsageCost` to `1/5000`. In other words, 5000 EVM gas would be counted as 1 Flow computation unit.
```

File: governance/20240508-computation-limit-hike.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+2. Reduce the coefficients of execution effort by a factor of 5x, effectively increasing the computation limit by the same factor. Counterbalance the reduction in coefficient values with a 5x increase in the cost of execution unit on Flow (from 4.99E-08 to 2.495E-7).
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:57]:
```suggestion
2. Reduce the coefficients of execution effort by a factor of 5, effectively increasing the computation limit by the same factor. 
3. Counterbalance the reduction in coefficient values with a 5x increase in the cost of execution unit on Flow (from 4.99E-08 to 2.495E-7).
```

File: governance/20240508-computation-limit-hike.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    **0.00234** * SetValue +
+		**8.65988** * CreateAccount +
+	  **1/5000** * EVMGasUsage**
+```
+
+## Impact
+
+The proposed/implemented changes aim to enhance the computation limit. For developers, the increase in effective computation limit along with the 5000:1 gas-to-compute ratio would empower them to execute larger contracts on EVM on Flow. The indirect method of increasing the compute limit (by changing the value of coefficients allowing larger transactions to “fit in”) eliminates any extra effort required to update their contracts for changes in the computation limit. Overall, this proposal would bring ‘EVM Equivalence’ in terms of contract and transaction size capabilities for EVM transactions on Flow.
+
+## Next Steps
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:72]: Can you throw in some examples - 
A token transfer would cost x in EVM gas and will use y in Cadence computation and will have a total transaction cost of Z FLOW.

File: governance/20240508-computation-limit-hike.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+To match Ethereum’s 30M gas limit, a 5000:1 gas to compute ratio could be adopted to calculate execution effort (compute units) on Flow. Based on some tests, this ratio would allow for 30M gas, along with adequate compute allocation for cadence execution of such transactions. For instance, a 30M gas transaction with a 5000:1 conversion ratio would consume about 6000 compute units for EVM execution, leaving 3999 compute units for Cadence execution.
+
+It’s important to note that for such sizable contracts however, even these compute units (3999 in the above example) may prove insufficient for cadence execution. Therefore, we should also consider revising Flow’s computation limit of 9999 to ensure seamless execution of the largest Ethereum contracts.
+
+1. **Computation Limit**
```
Comment (vishalchangrani) [inline governance/20240508-computation-limit-hike.md:46]:
```suggestion
 **Computation Limit**
```

File: governance/20240508-computation-limit-hike.md (hunk 4 above)
Comment (KshitijChaudhary666) [reply governance/20240508-computation-limit-hike.md:50]: Hmm its explained as an example in the next sentence. But lm try to change the language a bit!

File: governance/20240508-computation-limit-hike.md (hunk 8 above)
Comment (KshitijChaudhary666) [reply governance/20240508-computation-limit-hike.md:73]: This FLIP proposes no alterations that affect the fees so don't think we should talk about the costs here. The fee adjustments would only result from the 100x transaction fee FLIP proposal and also from introducing EVMGasUsageCost * EVMGasUsage to the execution effort calculation (explained in this post - https://forum.flow.com/t/how-evm-transaction-fees-work-on-flow-previewnet/5751). This FLIP is not gathering feedback on either of those so any examples to showcase cadence vs EVM fees might be confusing. Additionally, any examples (like the one in the post shared) would solely be illustrative at this point and cannot be generalized at all, which might add to the confusion

**Review by vishalchangrani — APPROVED** [review]

**Pull request merged**